src/mds/MDCache.cc

```diff
@@ -9713,10 +9713,18 @@ void MDCache::handle_discover(MDiscover *dis)
 
   assert(from != whoami);
 
+  if (mds->get_state() <= MDSMap::STATE_STARTING) {
+    dout(0) << "handle_discover: not started yet, delaying" << dendl;
+    mds->wait_for_mdsmap(mds->mdsmap->get_epoch(),
```

Comment:
mds should drop messages safely when its state < STATE_REJOIN. I think about this again, I think the root cause is that the sender does not re-send the discover message when the receiver mds become active. how about below patch.

<pre>
diff --git a/src/mds/MDCache.cc b/src/mds/MDCache.cc
index e81e1dc..1f2f664 100644
--- a/src/mds/MDCache.cc
+++ b/src/mds/MDCache.cc
@@ -9766,7 +9766,7 @@ void MDCache::handle_discover(MDiscover *dis)
 
   if (mds->get_state() <= MDSMap::STATE_REJOIN) {
     if (mds->get_state() < MDSMap::STATE_REJOIN &&
-       mds->get_want_state() != CEPH_MDS_STATE_REJOIN) {
+       mds->get_want_state() < CEPH_MDS_STATE_REJOIN) {
       dis->put();
       return;
     }
diff --git a/src/mds/MDSRank.cc b/src/mds/MDSRank.cc
index 10393af..f026ccb 100644
--- a/src/mds/MDSRank.cc
+++ b/src/mds/MDSRank.cc
@@ -1690,7 +1690,8 @@ void MDSRankDispatcher::handle_mds_map(
        oldmap->is_rejoining() && !mdsmap->is_rejoining())
       mdcache->dump_cache();      // for DEBUG only
 
-    if (oldstate >= MDSMap::STATE_REJOIN) {
+    if (oldstate >= MDSMap::STATE_REJOIN ||
+       oldstate == MDSMap::STATE_STARTING) {
       // ACTIVE|CLIENTREPLAY|REJOIN => we can discover from them.
       set<mds_rank_t> olddis, dis;
       oldmap->get_mds_set_lower_bound(olddis, MDSMap::STATE_REJOIN);
</pre>